openmls/src/codec/mod.rs

```diff
@@ -69,28 +69,50 @@ impl<'a> Cursor {
     }
 }
 
-pub trait Codec: Sized {
-    fn encode(&self, _buffer: &mut Vec<u8>) -> Result<(), CodecError> {
-        unimplemented!();
-    }
-
-    fn decode(_cursor: &mut Cursor) -> Result<Self, CodecError> {
-        unimplemented!();
-    }
+pub trait Encode: Sized {
+    fn encode(&self, _buffer: &mut Vec<u8>) -> Result<(), CodecError>;
 
     fn encode_detached(&self) -> Result<Vec<u8>, CodecError> {
         let mut buffer = vec![];
         self.encode(&mut buffer)?;
         Ok(buffer)
     }
+}
+
+pub trait Decode: Sized {
+    fn decode(_cursor: &mut Cursor) -> Result<Self, CodecError>;
 
     fn decode_detached(bytes: &[u8]) -> Result<Self, CodecError> {
         let cursor = &mut Cursor::new(bytes);
         Self::decode(cursor)
     }
 }
 
-impl Codec for u8 {
+pub trait Codec: Encode + Decode {}
+
+macro_rules! implement_codec {
```

Comment:
Is the macro really needed here? Can't we solve this with only using traits?

Reply:
This is a common macro use to safe writing out some code manually. As discussed I'll leave it in because it will go away soon anyway with the use of `tls_codec`.